com.aica.aicasystemclient/src/main/resources/com/aica/aicasystemclient/impl/daemon/server.py

```diff
@@ -0,0 +1,267 @@
+            res.close()
+        except urllib2.HTTPError as e:
+            raise RuntimeError('Failed to authenticate: ' + str(e))
+
+
+    @staticmethod
+    def _safe_uri(uri):
+        return urllib.quote_plus(uri)
+
+    def request(self, method, endpoint, params=None, json_data=None):
+        headers = {}
+        retry = 2
+        url = self._endpoint(endpoint)
+        if params:
+            url += '?' + urllib.urlencode(params)
+        data = None
+        if json_data is not None:
+            data = json_data and json.dumps(json_data).encode('utf-8')
+            headers['Content-Type'] = 'application/json'
+        while retry > 0:
+            if self.__api_key is not None:
+                self.__ensure_token()
+                headers['Authorization'] = 'Bearer ' + str(self.__token)
+            req = MethodRequest(url, data=data, headers=headers, method=method.upper())
+            try:
+                res = urllib2.urlopen(req, timeout=5)
+                response = Response(res.getcode(), res.read())
+                res.close()
+                return response
+            except urllib2.HTTPError as e:
+                if e.code == 401:
+                    if self.__api_key is None:
+                        break
+                    self.__token = None
+                else:
+                    return Response(e.code)
+            retry -= 1
+        return None
+    
+    def _check_version(self):
+        if self._core_version is None and self.core_version() is None:
+            return False
+        if self._core_version.startswith('4.') and self._core_version[2] in ['3', '4', '5']:
+            return True
+        return False
+
+    def core_version(self):
+        core_version = None
+        try:
+            res = urllib2.urlopen(self.__raw_endpoint('version'), timeout=5)
+            core_version = json.load(res)
+            res.close()
+        except urllib2.URLError:
+            pass
+        self._core_version = core_version
+        return self._core_version
+
+    def protocol(self):
+        protocol = None
+        try:
+            res = urllib2.urlopen(self.__raw_endpoint('protocol'), timeout=5)
+            protocol = json.load(res)
+            res.close()
+        except urllib2.URLError:
+            pass
+        self._protocol = protocol
+        return self._protocol
+
+    def check(self):
+        if self._protocol is None and self.protocol() is None:
+            return False
+        elif self._protocol != 'v2':
+            return False
+
+        if self._core_version is None and self.core_version() is None:
+            return False
+        
+        if self.__api_key is None:
+            return False
+        return True
+
+
+class Client:
+    def is_reachable(self):
+        return True
+
+    def initialize(self, url, api_key):
+        self.aica = AICA(url=url, api_key=api_key)
+        return self.aica.check()
+    
+    def decorator(func):
```

Comment:
nitpick: calling the decorator function `decorator` is a bit weird; even though it's not a reserved keyword it has a specific meaning in Python terms. It would be a bit like calling something `def function():`. I would prefer to call it `ensure_client` or something